discord/sink.py

```diff
@@ -0,0 +1,858 @@
+        written to file. The buffer allows leeway for packets that
+        arrive out of order to be reorganized.
+
+        Parameters
+        ----------
+        frame: :class:`AudioFrame`
+            The frame which will be added to the buffer.
+        """
+        if frame.ssrc not in self.output_files:
+            self.output_files[frame.ssrc] = open(os.path.join(self.output_dir, f"audio-{frame.ssrc}.pcm"), "wb")
+            self._frame_buffer[frame.ssrc] = []
+        self._frame_buffer[frame.ssrc].append(frame)
+        if len(self._frame_buffer[frame.ssrc]) >= self.FRAME_BUFFER_LIMIT:
+            self._write_buffer(frame.ssrc)
+
+    def on_rtcp(self, packet: RTCPPacket) -> None:
+        """This function receives RTCP Packets, but does nothing with them since
+        there is no use for them in this sink.
+
+        Parameters
+        ----------
+        packet: :class:`RTCPPacket`
+            A RTCP Packet received from discord. Can be any of the following:
+            :class:`RTCPSenderReportPacket`, :class:`RTCPReceiverReportPacket`,
+            :class:`RTCPSourceDescriptionPacket`, :class:`RTCPGoodbyePacket`,
+            :class:`RTCPApplicationDefinedPacket`
+        """
+        pass
+
+    def _write_buffer(self, ssrc, empty=False):
+        self._frame_buffer[ssrc] = sorted(self._frame_buffer[ssrc], key=lambda frame: frame.sequence)
+        index = self.FRAME_BUFFER_LIMIT // 2 if not empty else self.FRAME_BUFFER_LIMIT
+        for frame in self._frame_buffer[ssrc][:index]:
+            self._write_frame(frame)
+        self._frame_buffer[ssrc] = self._frame_buffer[ssrc][index:]
+
+    def _write_frame(self, frame):
+        if frame.ssrc in self._timestamps:
+            # write silence
+            silence = frame.timestamp - self._timestamps[frame.ssrc] - OpusDecoder.FRAME_SIZE
+            if silence > 0:
+                self.output_files[frame.ssrc].write(b"\x00" * silence * OpusDecoder.CHANNELS)
+        self.output_files[frame.ssrc].write(frame.audio)
+        self._timestamps[frame.ssrc] = frame.timestamp
+        self._ssrc_to_user[frame.ssrc] = frame.user
+
+    def cleanup(self) -> None:
+        """Writes remaining frames in buffer and closes all files"""
+        if self.done:
+            return
+        for ssrc in self._frame_buffer.keys():
+            self._write_buffer(ssrc, empty=True)
+        for file in self.output_files.values():
+            file.close()
+        self.done = True
+
+    def convert_files(self) -> None:
+        """Goes through all files, opens them and calls convert_file, then deletes them.
+
+        convert_file must be defined, else NotImplementedError will be raised
+        """
+        if not self.done:
+            self.cleanup()
+        for ssrc, file in self.output_files.items():
+            filepath = file.name
+            user = self._ssrc_to_user[ssrc]
+            new_name = f"audio-{user.name}#{user.discriminator}-{ssrc}" if user is not None else None
```

Comment:
new_name = f"audio-{user}-{ssrc}" if user else None

Reply:
finally it seems that use is:
- the identifier
- or the object discord.py
I search why.

Reply:
Line #680 is the reason. Since AudioFrame.user can be int or Member, it doesn't check first.

Had this PR been reviewed with an active TypeChecker this would have been caught.

Reply:
Implemented a function that decides the file name format based on the type of `user`